README.md
# pcexpress-pickup
A script to check for available pickup times at PC Express locations

## Usage
```
$ python check_all_by_postal_code.py
usage: check_all_by_postal_code.py [-h] [-p P] [-lat LAT] [-long LONG] [-d D] [-r] [-id ID] [-brands BRANDS]

optional arguments:
  -h, --help         show this help message and exit
  -p P, -postal P    ex: m5w1e6
  -lat LAT           ex: 43.703344
  -long LONG         ex: -79.524619
  -d D, -distance D  Search distance in KM
  -r, -report        report lat + long and stores found within search distance. will not check available pickup times
  -id ID             comma seperated store IDs ex: 1111,1122
  -brands BRANDS     filter by comma seperated store brands ex: loblaw,superstore,zehrs
```

### Report a list of stores available within a 10KM radius
```
$ python check_all_by_postal_code.py -p m5e1w6 -d 10 -r
For faster lookup times use: -lat 43.703344 -long -79.524619
superstore, id: 2800, 2549 Weston Rd Toronto, Ontario M9N 2A7, approx 0.9 KM away
<...>
fortinos, id: 0096, 3940 Hwy 7 RR 2 Vaughan, Ontario L4L 9C3, approx 9.8 KM away
```

### Use -lat and -long to avoid doing a postal code lookup each time
```
$ python check_all_by_postal_code.py -lat 43.703344 -long -79.524619 -d 10 
9 available at nofrills at 245 Dixon Rd Etobicoke, Ontario M9P 2M4 approx 1.8 KM away
2020-04-14 09:00:00-04:00
<...>
2020-04-23 19:00:00-04:00

1 available at loblaw at 270 The Kingsway Etobicoke, Ontario M9A 3T7 approx 4.7 KM away
2020-04-23 16:00:00-04:00

1 available at loblaw at 3671 Dundas St W Toronto, Ontario M6S 2T3 approx 4.8 KM away
2020-04-23 16:00:00-04:00
```
### Use -id to search for pickup times at specific stores. Geo lookup disaabled, so distance will always be 0 KM
```
$ python check_all_by_postal_code.py -id 2800,3480,1174
20 available at loblaw at 270 The Kingsway Etobicoke, Ontario M9A 3T7 approx 0 KM away
2020-04-14 14:00:00-04:00
<...>
2020-04-23 15:00:00-04:00
```
## Installation using a python3 virtual environment - python 3.6 or newer required
```bash
$ git clone https://github.com/shmick/pcexpress-pickup
$ python3 -m venv pcexpress-pickup/
$ cd pcexpress-pickup/
$ source bin/activate
$ pip install -r requirements.txt 
```

## Run via Docker
```bash
$ docker run shmick/pcexpress-pickup -lat 43.703344 -long -79.524619
8 available at nofrills at 245 Dixon Rd Etobicoke, Ontario M9P 2M4 approx 1.8 KM away
2020-04-23 12:00:00-04:00
<...>
2020-04-23 19:00:00-04:00

1 available at loblaw at 3671 Dundas St W Toronto, Ontario M6S 2T3 approx 4.8 KM away
2020-04-23 16:00:00-04:00
```

## Build and run your own container
```bash
$ git clone https://github.com/shmick/pcexpress-pickup
$ cd pcexpress-pickup/
$ docker build -t pcexpress-pickup:latest .
$ docker run shmick/pcexpress-pickup
```

# Notes
https://www.pcexpress.ca/bundle.js contains a list of all pcexpress pickup locations as well as plenty of other store and location metadata. 

Active stores locations have been extracted to locations.json, which is what the utility uses.

The entire unfiltered location dataset has been saved to unfiltered-locations.json

### location sample data
```json
{
  "locations": [
    {
      "id": "0925",
      "contactNumber": "7096430850",
      "name": "Prince Rupert Street",
      "storeId": "0925",
      "storeBannerId": "dominion",
      "locationType": "STORE",
      "pickupType": "STORE",
      "bufferTimeInHours": 2,
      "partner": null,
      "visible": true,
      "isShoppable": true,
      "geoPoint": {
        "latitude": 48.546052,
        "longitude": -58.582397
      },
      "address": {
        "country": "Canada",
        "region": "Newfoundland and Labrador",
        "town": "Stephenville",
        "line1": "62 Prince Rupert St",
        "line2": null,
        "postalCode": "A2N 3W7",
        "formattedAddress": "62 Prince Rupert St Stephenville, Newfoundland and Labrador A2N 3W7"
      },
      "timeZone": "Canada/Newfoundland",
      "features": []
    },
    {
      "id": "0927",
      "contactNumber": "7096513437",
      "name": "Laurell Road",
      "storeId": "0927",
      "storeBannerId": "dominion",
      "locationType": "STORE",
      "pickupType": "STORE",
      "bufferTimeInHours": 2,
      "partner": null,
      "visible": true,
      "isShoppable": true,
      "geoPoint": {
        "latitude": 48.949657,
        "longitude": -54.60071
      },
      "address": {
        "country": "Canada",
        "region": "Newfoundland and Labrador",
        "town": "Gander",
        "line1": "100 Laurell Rd",
        "line2": null,
        "postalCode": "A1V 2V5",
        "formattedAddress": "100 Laurell Rd Gander, Newfoundland and Labrador A1V 2V5"
      },
      "timeZone": "Canada/Newfoundland",
      "features": []
    }
  ]
}
```

CHANGELOG.md
## 2020-04-19

* remove unnecessary for loop when checking store IDs and brands

* dedupe list of store IDs and brands 

* alter internals to use lists for storing found timeslots 

* cleanup __main__ entry 

* fix filter by brands logic. If the brand wasn't in the results, it would show the full results.

## 2020-04-18

* implement generic cvs string to list function 

* use timezone of system running the script instead of fixed to America/Toronto

* add -brands option to filter result by store brand (superstore,loblaw,...)

```
usage: check_all_by_postal_code.py [-h] [-p P] [-lat LAT] [-long LONG] [-d D] [-r] [-id ID] [-brands BRANDS]

optional arguments:
  -h, --help         show this help message and exit
  -p P, -postal P    ex: m5w1e6
  -lat LAT           ex: 43.703344
  -long LONG         ex: -79.524619
  -d D, -distance D  Search distance in KM
  -r, -report        report lat + long and stores found within search distance. will not check available pickup times
  -id ID             comma seperated store IDs ex: 1111,1122
  -brands BRANDS     filter by comma seperated store brands ex: loblaw,superstore,zehrs
```

## 2020-04-11

* split into task specific functions

## 2020-04-10

** BREAKING CHANGES **

* add -id option to search by store IDs

* add argparse to enable proper command line switches

```
usage: check_all_by_postal_code.py [-h] [-p P] [-lat LAT] [-long LONG] [-d D] [-r] [-id ID]

optional arguments:
  -h, --help         show this help message and exit
  -p P, -postal P    ex: m5w1e6
  -lat LAT           ex: 43.703344
  -long LONG         ex: -79.524619
  -d D, -distance D  Search distance in KM
  -r, -report        report lat + long and stores found within search distance. will not check available pickup times
  -id ID             comma seperated store IDs ex: 1111,1122
```
* remove MYLAT and MYLONG environment variables

* prefer `-lat` and `-long` over `-p` postal code

## 2020-04-09

* check for MYLAT and MYLONG environment variables to bypass geo lookup request. If the report option is used, it will give you the variables to use along with the locations.
```bash
$ python check_all_by_postal_code.py m5e1w6 2 report
export MYLAT="43.703344" ; export MYLONG="-79.524619"
store: superstore, location: 2800, address: 2549 Weston Rd Toronto, Ontario M9N 2A7, approx 0 KM away
store: nofrills, location: 3480, address: 245 Dixon Rd Etobicoke, Ontario M9P 2M4, approx 1 KM away
```
* geo lookup throttling fix

* renamed id variable to loc_id as `id()` is a built-in python function

* additional filtering of locations.json

## 2020-04-08

* fallback to NRC geo lookup in the event that geocode.ca throttles requests

* switch to geocode.ca as primary geo lookup for more reliable results

* store locations are displayed in order of distance closest to you 

* add secondary postal code lookup since the NRC geo lookup may return nothing

## 2020-04-06

* add Dockerfile and link to dockerhub

* add `report` mode to check_all_by_postal_code.py

* add postal_code variable

* check to ensure a minimum python version of 3.6 is being used

* change base_url to a dictionary lookup

## 2020-04-05

* merge [Lookup by PostCode / Specify Max Dist in KM feature](https://github.com/shmick/pcexpress-pickup/pull/1) ( Big thanks to [adam1](https://github.com/1adam) )

## 2020-04-01

* add LICENSE

* add stores.json and update locations.json

* initial commit

test/tests.py
from check_all_by_postal_code import get_ids, get_latlog
import unittest


class Tests(unittest.TestCase):
    def test_ids(self):
        result = get_ids("1111,2222")
        self.assertEqual(result, ["1111", "2222"])

    def test_get_latlog_supplied(self):
        myLat = 22
        myLong = 33
        report = False
        postal_code = None
        result = get_latlog(postal_code, myLat, myLong, report)
        self.assertEqual(result, (22, 33))

    def test_get_latlog_postal(self):
        myLat = None
        myLong = None
        report = False
        postal_code = "m5w1e6"
        result = get_latlog(postal_code, myLat, myLong, report)
        self.assertEqual(result, (43.683428, -79.392511))


if __name__ == "__main__":
    unittest.main()


check.py
import requests
from datetime import datetime
from dateutil import tz

from_zone = tz.gettz("UTC")
to_zone = tz.gettz("America/Toronto")

loblaws_dict = {
    "1208": "10909 Yonge St Richmond Hill, Ontario L4C 3E3",
    "1028": "301 High Tech Rd Richmond Hill, Ontario L4B 4R2",
}

rcss_dict = {
    "1030": "15900 Bayview Ave Aurora, Ontario L4G 7Y3",
}

nofrills_dict = {
    "0770": "9325 Yonge St Richmond Hill, Ontario L4C 0A8",
    "3652": "14800 Yonge St #162 Aurora, Ontario L4G 1N3",
}


def check_loblaws(storeBannerId, id):

    if storeBannerId == "loblaws":
        base_url = "https://www.loblaws.ca"
        # If this header isn't set, the site returns an error
        headers = {"Site-Banner": "loblaw"}
        address = loblaws_dict[id]

    if storeBannerId == "rcss":
        base_url = "https://www.realcanadiansuperstore.ca"
        # If this header isn't set, the site returns an error
        headers = {"Site-Banner": "superstore"}
        address = rcss_dict[id]

    if storeBannerId == "nofrills":
        base_url = "https://www.nofrills.ca"
        # If this header isn't set, the site returns an error
        headers = {"Site-Banner": "nofrills"}
        address = nofrills_dict[id]

    # Using the base_url and headers, build the full URL
    url = f"{base_url}/api/pickup-locations/{id}/time-slots"
    # Make the HTTP request
    r = requests.get(url, headers=headers)
    # Use the builtin JSON decoder
    data = r.json()
    # We only want to process the timeSlots entries from the output
    timeslots = data["timeSlots"]
    # Initialze an empty string to store the pickup_times
    pickup_times = ""

    # Loop through the results
    for startTime in timeslots:
        # Get a list of pickup times where the "available" value is not False
        if not startTime.get("available") is False:
            # Convert the UTC times to local time
            start_time = local_time(startTime["startTime"])
            # Append the converted start_time to the results
            pickup_times += start_time
            pickup_times += "\n"

    if pickup_times:
        count = pickup_times.count("\n")
        output = f"{count} pickup times available at {storeBannerId} at {address}"
        output += "\n"
        output += pickup_times

        print(output)


# Convert the UTC timestamps to localtime
def local_time(timestamp):
    utc = datetime.strptime(timestamp, "%Y-%m-%dT%H:%M:%SZ")
    utc = utc.replace(tzinfo=from_zone)
    local = utc.astimezone(to_zone)
    return str(local)


check_loblaws("loblaws", "1208")
check_loblaws("loblaws", "1028")
check_loblaws("rcss", "1030")
check_loblaws("nofrills", "0770")
check_loblaws("nofrills", "3652")


check_all_by_postal_code.py
from __future__ import with_statement
import os
import sys
import json
import argparse
import requests
import haversine
import datetime as dt
from datetime import datetime
from dateutil import tz
from haversine import haversine, Unit


MIN_PYTHON = (3, 6)  # Python 3.6
args = sys.argv[1:]
geo_pri = "https://geocoder.ca/?geoit=xml&json=1&postal="
geo_bak = "https://geogratis.gc.ca/services/geolocation/en/locate?q="
store_locations = "./locations.json"


def parse_args(args):
    parser = argparse.ArgumentParser()

    parser.add_argument("-p", "-postal", type=str, help="ex: m5w1e6")
    parser.add_argument("-lat", help="ex: 43.703344", type=float)
    parser.add_argument("-long", help="ex: -79.524619", type=float)
    parser.add_argument(
        "-d", "-distance", help="Search distance in KM", default=5.0, type=float,
    )
    parser.add_argument(
        "-r",
        "-report",
        action="store_true",
        help="report lat + long and stores found within search distance. will not check available pickup times",
    )
    parser.add_argument("-id", type=str, help="comma seperated store IDs ex: 1111,1122")
    parser.add_argument(
        "-brands",
        type=str,
        help="filter by comma seperated store brands ex: loblaw,superstore,zehrs",
    )
    parser.add_argument(
        "--verbose", action="store_true", help=argparse.SUPPRESS,
    )
    args = parser.parse_args(args)
    if len(sys.argv) == 1:
        parser.print_help(sys.stderr)
        sys.exit(1)
    return args


def string_to_list(csv_string):
    deduped = []
    stringlist = []
    if csv_string:
        stringlist = csv_string.split(",")
    for entry in stringlist:
        if entry not in deduped:
            deduped.append(entry)
    return deduped


def check_params(postal_code, myLat, myLong):
    if not postal_code and not myLat and not myLong:
        sys.exit(
            "Either postal code (-p) or latitude (-lat) and longitude (-long) values are required. Use -h for help"
        )

    if not postal_code:
        if not myLat or not myLong:
            sys.exit(
                "Both latitude (-lat) and longitude (-long) values are required. Use -h for help"
            )


def geo_lookup(postal_code):
    geo_url_pri = geo_pri + postal_code
    try:
        geo_data = requests.get(geo_url_pri)
    except requests.exceptions.RequestException as e:
        raise SystemExit(e)
    # geocoder.ca can throttle requests. In this case, use the backup geolocation lookup.
    if geo_data.status_code == 200:
        geo_json = geo_data.json()
        myLat = float(geo_json["latt"])
        myLong = float(geo_json["longt"])
    else:
        geo_url_bak = geo_bak + postal_code
        try:
            geo_data = requests.get(geo_url_bak)
        except requests.exceptions.RequestException as e:
            raise SystemExit(e)
        geo_json = geo_data.json()
        coords = geo_json[0]["geometry"]["coordinates"]
        myLat = coords[1]
        myLong = coords[0]
    myLatLong = (myLat, myLong)
    return myLatLong


def get_latlog(postal_code, myLat, myLong, report):
    myLatLong = None
    if postal_code:
        if not myLat or not myLong:
            myLatLong = geo_lookup(postal_code)
    if myLat and myLong:
        myLatLong = (myLat, myLong)
    if report and not my_ids:
        print(f"For faster lookup times use: -lat {myLatLong[0]} -long {myLatLong[1]}")
    return myLatLong


def check_stores_by_id(my_ids, all_locs):
    stores_to_check = []
    for loc in all_locs["locations"]:
        store_id = loc["id"]
        if store_id in my_ids:
            # when using -id the distance to the store will be 0
            loc_distance = {"distance": 0}
            loc.update(loc_distance)
            stores_to_check.append(loc)
    return stores_to_check


def check_stores_by_geo(filter_brands, myLatLong, all_locs, within_km):
    stores_to_check = []
    stores_to_check_filtered_by_brand = []
    for loc in all_locs["locations"]:
        locLatLong = (loc["geoPoint"]["latitude"], loc["geoPoint"]["longitude"])
        distance = haversine(myLatLong, locLatLong)
        if distance <= within_km:
            # Add a distance key:value pair to the location object so that
            # locations can be sorted by distance after the list is created
            loc_distance = {"distance": distance}
            loc.update(loc_distance)
            stores_to_check.append(loc)
    if filter_brands:
        if stores_to_check:
            for store in stores_to_check:
                storeBannerId = store["storeBannerId"]
                if storeBannerId in filter_brands:
                    stores_to_check_filtered_by_brand.append(store)
        return stores_to_check_filtered_by_brand
    else:
        return stores_to_check


def stores_to_check(store_locations, filter_brands, my_ids):
    try:
        with open(store_locations) as inf:
            all_locs = json.load(inf)
    except EnvironmentError as e:  # parent of IOError, OSError *and* WindowsError where available
        sys.exit(e)

    filter_brands = string_to_list(filter_brands)
    my_ids = string_to_list(my_ids)

    stores_to_check = []

    # if store IDs are specified, no need to figure out stores based on distance
    if my_ids:
        stores_to_check = check_stores_by_id(my_ids, all_locs)
    else:
        check_params(postal_code, myLat, myLong)
        myLatLong = get_latlog(postal_code, myLat, myLong, report)
        stores_to_check = check_stores_by_geo(
            filter_brands, myLatLong, all_locs, within_km
        )

    if stores_to_check:
        # Sort the list of locations by distance
        stores_to_check = sorted(stores_to_check, key=lambda s: s["distance"])
    return stores_to_check


def check_loblaws(store):
    loc_id = store["id"]
    address = store["address"]["formattedAddress"]
    storeBannerId = store["storeBannerId"]
    distance = round(store["distance"], 1)

    output_data = []

    if report:
        output_data.append(
            f"{storeBannerId}, id: {loc_id}, {address}, approx {distance} KM away"
        )
    else:
        if verbose:
            print(
                f"checking {storeBannerId}, id: {loc_id}, {address}, approx {distance} KM away"
            )
        base_url = store_urls[storeBannerId]
        headers = {
            "Site-Banner": storeBannerId
        }  # If this header isn't set, the site returns an error
        url = base_url + "/api/pickup-locations/" + loc_id + "/time-slots"
        try:
            r = requests.get(url, headers=headers)
        except requests.exceptions.RequestException as e:
            raise SystemExit(e)
        # Use the builtin JSON decoder
        data = r.json()
        # We only want to process the timeSlots entries from the output
        try:
            timeslots = data["timeSlots"]
        except:
            return
        # Loop through the results
        for startTime in timeslots:
            # Get a list of pickup times where the "available" value is not False
            if not startTime.get("available") is False:
                # Convert the UTC times to local time
                startTime = local_time(startTime["startTime"])
                # Append the converted start_time to the results
                output_data.append(startTime)
        if output_data:
            count = len(output_data)
            # Insert the count and store info at the beginning of the list
            output_data.insert(
                0,
                f"{count} available at {storeBannerId} at {address} approx {distance} KM away",
            )
            # Add an empty entry to the list to create an empty line between each store output
            output_data.append("")
    return output_data


# Convert the UTC timestamps to localtime
from_zone = tz.gettz("UTC")
# to_zone = tz.gettz("America/Toronto")
# Rather than hard code the timezone, figure out the local timezone
to_zone = dt.datetime.now(dt.timezone(dt.timedelta(0))).astimezone().tzinfo


def local_time(timestamp):
    utc = datetime.strptime(timestamp, "%Y-%m-%dT%H:%M:%SZ")
    utc = utc.replace(tzinfo=from_zone)
    local = utc.astimezone(to_zone)
    return str(local)


store_urls = {
    "bloor": "https://www.bloorstreetmarket.ca",
    "box": "https://www.boxfoodstores.ca",
    "independent": "https://www.yourindependentgrocer.ca",
    "dominion": "https://www.newfoundlandgrocerystores.ca",
    "extra": "https://www.extrafoods.ca",
    "fortinos": "https://www.fortinos.ca",
    "loblaw": "https://www.loblaws.ca",
    "independentcitymarket": "https://www.independentcitymarket.ca",
    "provigo": "https://www.provigo.ca",
    "rass": "https://www.atlanticsuperstore.ca",
    "saveeasy": "https://www.saveeasy.ca",
    "superstore": "https://www.superstore.ca",
    "valumart": "https://www.valumart.ca",
    "zehrs": "https://www.zehrs.ca",
    "maxi": "https://www.maxi.ca",
    "nofrills": "https://www.nofrills.ca",
    "wholesaleclub": "https://www.wholesaleclub.ca",
    "tntsupermarket": "https://www.tntsupermarket.com",
}


def check_python_ver(info):
    if info < MIN_PYTHON:
        sys.exit("Python %s.%s or later is required.\n" % MIN_PYTHON)


if __name__ == "__main__":
    check_python_ver(sys.version_info)
    args = parse_args(args)
    postal_code = args.p
    myLat = args.lat
    myLong = args.long
    within_km = args.d
    report = args.r
    my_ids = args.id
    filter_brands = args.brands
    verbose = args.verbose

    if verbose:
        print(args)

    # build the list of store locations to check
    stores_to_check = stores_to_check(store_locations, filter_brands, my_ids)
    if stores_to_check:
        for store in stores_to_check:
            result = check_loblaws(store)
            if result:
                for entries in result:
                    print(entries)
